common/src/main/java/com/microsoft/identity/common/internal/platform/DevicePopManager.java

```diff
@@ -0,0 +1,834 @@
+        }
+
+        // Clean up... we generated a cert, but it cannot be used.
+        clearAsymmetricKey();
+
+        throw new UnsupportedOperationException(
+                "Failed to generate valid KeyPair. Attempted " + MAX_RETRIES + " times."
+        );
+    }
+
+    private void logSecureHardwareState(@NonNull final KeyPair kp) {
+        String msg;
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            try {
+                final PrivateKey privateKey = kp.getPrivate();
+                final KeyFactory factory = KeyFactory.getInstance(
+                        privateKey.getAlgorithm(), ANDROID_KEYSTORE
+                );
+                final KeyInfo info = factory.getKeySpec(privateKey, KeyInfo.class);
+                final boolean isInsideSecureHardware = info.isInsideSecureHardware();
+                msg = "SecretKey is secure hardware backed? " + isInsideSecureHardware;
+            } catch (final Exception e) {
+                msg = "Failed to query secure hardware state.";
+            }
+        } else {
+            msg = "Cannot query secure hardware state (API unavailable <23)";
+        }
+
+        Logger.info(TAG, msg);
+    }
+
+    /**
+     * Generates a new {@link KeyPair}.
+     *
+     * @param ctx          The application Context.
+     * @param useStrongbox True if StrongBox should be used, false otherwise.
+     * @return The newly generated KeyPair.
+     * @throws InvalidAlgorithmParameterException If the designated crypto algorithm is not
+     *                                            supported for the designated parameters.
+     * @throws NoSuchAlgorithmException           If the designated crypto algorithm is not supported.
+     * @throws NoSuchProviderException            If the designated crypto provider cannot be found.
+     * @throws StrongBoxUnavailableException      If StrongBox is unavailable.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
+    private KeyPair generateNewKeyPair(@NonNull final Context ctx, final boolean useStrongbox)
+            throws InvalidAlgorithmParameterException, NoSuchAlgorithmException,
+            NoSuchProviderException, StrongBoxUnavailableException {
+        final KeyPairGenerator kpg = getInitializedRsaKeyPairGenerator(
+                ctx,
+                RSA_KEY_SIZE,
+                useStrongbox
+        );
+        return kpg.generateKeyPair();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
+    private KeyPairGenerator getInitializedRsaKeyPairGenerator(@NonNull final Context ctx,
+                                                               final int keySize,
+                                                               final boolean useStrongbox)
+            throws InvalidAlgorithmParameterException, NoSuchProviderException, NoSuchAlgorithmException {
+        // Create the KeyPairGenerator
+        final KeyPairGenerator keyPairGenerator = getKeyPairGenerator(
```

Comment:
nit: final KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(KeyPairGeneratorAlgorithms.RSA, ANDROID_KEYSTORE)

Reply:
Resolved in 5bc2da5d9dc75f3624285052c3f1c7ffb6a561e5